File: src/Aiko.Core/PersonaPlugin.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Nodes;

namespace Aiko.Core;

/// The files of the aiko-persona plugin. The bridge writes them into a folder and prints its path,
/// and Claude Code copies that folder into its plugin cache (D-201). The version Claude Code keeps
/// is a hash of this content, so nothing here counts versions by hand.
public static class PersonaPlugin
{
    public const string Name = "aiko-persona";

    /// The hook argument the bridge answers to (D-206).
    public const string HookArgument = "hook";

    /// The events that change Aiko's face (D-206). Notification is filtered by the bridge, which
    /// keeps only the kinds that mean "waiting for you".
    public static readonly IReadOnlyList<string> HookEvents =
    [
        "UserPromptSubmit", "PostToolUse", "PermissionRequest", "Notification", "Stop", "StopFailure", "SessionEnd",
    ];

    /// Relative path with forward slashes → file content.
    public static IReadOnlyDictionary<string, string> Files(Temperament temperament, string bridgeExePath) =>
        new SortedDictionary<string, string>(StringComparer.Ordinal)
        {
            [".claude-plugin/plugin.json"] = Manifest(),
            ["output-styles/aiko.md"] = PersonaPrompt.StyleFile(temperament),
            ["hooks/hooks.json"] = Hooks(bridgeExePath),
        };

    /// Twelve hex characters of SHA-256 over the paths and contents. The bridge names the folder
    /// after it, so a changed persona lands in a new folder and never half-overwrites the old one.
    public static string ContentHash(IReadOnlyDictionary<string, string> files)
    {
        using var sha = SHA256.Create();
        foreach (var (path, content) in files.OrderBy(f => f.Key, StringComparer.Ordinal))
        {
            var bytes = Encoding.UTF8.GetBytes($"{path}\n{content}\n");
            sha.TransformBlock(bytes, 0, bytes.Length, null, 0);
        }

        sha.TransformFinalBlock([], 0, 0);
        return Convert.ToHexStringLower(sha.Hash!)[..12];
    }

    private static string Manifest() =>
        new JsonObject
        {
            ["name"] = Name,
            ["description"] = "Aiko's persona and the hooks behind her face in the tray.",
        }.ToJsonString(Indented) + "\n";

    /// The command and its argument go separately, so no shell reads the path: the install path
    /// has spaces and may have letters that are not ASCII. async keeps the session from waiting.
    private static string Hooks(string bridgeExePath)
    {
        var events = new JsonObject();
        foreach (var name in HookEvents)
        {
            events[name] = new JsonArray(
                new JsonObject
                {
                    ["hooks"] = new JsonArray(
                        new JsonObject
                        {
                            ["type"] = "command",
                            ["command"] = bridgeExePath,
                            ["args"] = new JsonArray(HookArgument),
                            ["async"] = true,
                        }),
                });
        }

        return new JsonObject { ["hooks"] = events }.ToJsonString(Indented) + "\n";
    }

    private static readonly System.Text.Json.JsonSerializerOptions Indented = new()
    {
        WriteIndented = true,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };
}

File: mac/Sources/AikoKit/PersonaPlugin.swift
import CryptoKit
import Foundation

/// The files of the aiko-persona plugin. The bridge writes them into a folder and prints its path,
/// and Claude Code copies that folder into its plugin cache (D-201). The version Claude Code keeps
/// is a hash of this content, so nothing here counts versions by hand.
public enum PersonaPlugin {
    public static let name = "aiko-persona"

    /// The hook argument the bridge answers to (D-206).
    public static let hookArgument = "hook"

    /// The events that change Aiko's face (D-206). Notification is filtered by the bridge, which
    /// keeps only the kinds that mean "waiting for you".
    public static let hookEvents = [
        "UserPromptSubmit", "PostToolUse", "PermissionRequest", "Notification", "Stop", "StopFailure", "SessionEnd",
    ]

    /// Relative path with forward slashes to file content, in the order the paths sort.
    public static func files(_ temperament: Temperament, _ bridgePath: String) -> [(path: String, content: String)] {
        [
            (".claude-plugin/plugin.json", manifest()),
            ("hooks/hooks.json", hooks(bridgePath)),
            ("output-styles/aiko.md", PersonaPrompt.styleFile(temperament)),
        ]
    }

    /// Twelve hex characters of SHA-256 over the paths and contents. The bridge names the folder
    /// after it, so a changed persona lands in a new folder and never half-overwrites the old one.
    public static func contentHash(_ files: [(path: String, content: String)]) -> String {
        var sha = SHA256()
        for file in files.sorted(by: { $0.path < $1.path }) {
            sha.update(data: Data("\(file.path)\n\(file.content)\n".utf8))
        }

        let hex = sha.finalize().map { String(format: "%02x", $0) }.joined()
        return String(hex.prefix(12))
    }

    private static func manifest() -> String {
        let object = JsonObject([
            ("name", .string(name)),
            ("description", .string("Aiko's persona and the hooks behind her face in the tray.")),
        ])
        return JsonNode.object(object).toJsonString(indented: true, escaping: .relaxed) + "\n"
    }

    /// The command and its argument go separately, so no shell reads the path: the install path
    /// has spaces and may have letters that are not ASCII. async keeps the session from waiting.
    private static func hooks(_ bridgePath: String) -> String {
        var events = JsonObject()
        for name in hookEvents {
            events[name] = .array([
                .object(JsonObject([
                    ("hooks", .array([
                        .object(JsonObject([
                            ("type", .string("command")),
                            ("command", .string(bridgePath)),
                            ("args", .array([.string(hookArgument)])),
                            ("async", .bool(true)),
                        ]))
                    ]))
                ]))
            ])
        }

        let root = JsonObject([("hooks", .object(events))])
        return JsonNode.object(root).toJsonString(indented: true, escaping: .relaxed) + "\n"
    }
}
